tabular/src/autogluon/tabular/models/tab_transformer/tab_transformer_model.py

```diff
@@ -0,0 +1,469 @@
+        else:
+            train.targets = torch.LongTensor(list(y_train))
+            val.targets = torch.LongTensor(list(y_val))
+
+        trainloader, valloader, unlabloader = train.build_loader(
+            shuffle=True), val.build_loader(), unlab.build_loader() if unlab is not None else None
+
+        self.cat_feat_origin_cards = trainloader.cat_feat_origin_cards
+
+        self.get_model()
+
+        if X_unlabeled is not None:
+            self.tt_fit(unlabloader, valloader, y_val, state='pretrain', time_limit=time_limit)
+            self.tt_fit(trainloader, valloader, y_val, state='finetune', time_limit=time_limit)
+        else:
+            self.tt_fit(trainloader, valloader, y_val, time_limit=time_limit)
+
+    def _predict_proba(self, X, preprocess=False):
+        """
+        X (torch.tensor or pd.dataframe): data for model to give prediction probabilities
+        returns: np.array of k-probabilities for each of the k classes. If k=2 we drop the second probability.
+        """
+        import torch
+        import torch.nn as nn
+        from torch.utils.data import DataLoader
+        from torch.autograd import Variable
+
+        if isinstance(X, pd.DataFrame):
+            if preprocess:
+                X = self.preprocess(X)
+            # Internal preprocessing, renaming col names, tt specific.
+            X, _, _ = self._tt_preprocess(X, fe=self.fe)
+            loader = X.build_loader()
+        elif isinstance(X, DataLoader):
+            loader = X
+        elif isinstance(X, torch.Tensor):
+            X = X.rename(columns=self._get_no_period_columns(X))
+            loader = X.build_loader()
+        else:
+            raise NotImplementedError(
+                "Attempting to predict against a non-supported data type. \nNeeds to be a pandas DataFrame, torch DataLoader or torch Tensor.")
+
+        self.model.eval()
+        softmax = nn.Softmax(dim=1)
+
+        if self.problem_type == REGRESSION:
+            outputs = torch.zeros([len(loader.dataset), 1])
+        else:
+            outputs = torch.zeros([len(loader.dataset), self.num_classes])
+
+        iter = 0
+        for data, _ in loader:
+            if self.params['device'].type == "cuda":
+                data = data.cuda()
+            with torch.no_grad():
+                data = Variable(data)
+                prob, _ = self.model(data)
+                batch_size = len(prob)
+                if self.problem_type != REGRESSION:
+                    prob = softmax(prob)
+
+            outputs[iter:(iter + batch_size)] = prob
+            iter += batch_size
+
+        if self.problem_type == BINARY:
+            return outputs[:, 1].cpu().numpy()
+        elif self.problem_type == REGRESSION:
+            outputs = outputs.flatten()
+            #return outputs.flatten().numpy()
+
+        return outputs.cpu().numpy()
+
+    def _get_default_searchspace(self):
+        return get_default_searchspace()
+
+
+    # # TODO: Do I need X_unlabeled here?
```

Comment:
can remove. Just leave TODO to consider HPO for pretraining with unlabeled data.

Reply:
Changed TODO to be appropriate for future work.